### ISSUE TITLE
Non-Latin alphabet support

### ISSUE THREAD

**Issue by ystyle** [gabm/Satty#2]
Unable to activate fcitx5 input method

**Comment by gabm** [comment]
thank you for this. Can you provide more information in what this is and what is useful for?

if you are a developer: do you have resources that demonstrate it?

**Comment by ystyle** [comment]
Add multiple input methods to fcitx5, use `ctrl + space` to switch input methods when annotating, but the input method does not respond.
I have never learned rust. Neovide solved this problem in x11 not long ago.

**Comment by lilydjwg** [comment]
If you can't put a GtkEntry there, you'll need to interact using [`GtkIMContext`](https://docs.gtk.org/gtk4/class.IMContext.html) yourself.

**Comment by gabm** [comment]
i really didn't get what you wanted to do as the interface is mainly pointer based - but you do mean the text box input, right?

**Comment by lilydjwg** [comment]
Yes. That's the only place where the user can input text, I suppose?

Basically, you need to create a GtkIMContext, call `gtk_im_context_focus_in` when the user is going to input text, and call `gtk_im_context_set_cursor_location` when the text cursor moves, listen to `commit` signal to get the inputted text and insert it to your text buffer. You also need to handle those `preedit-{start,changed,end}` signals and use `gtk_im_context_get_preedit_string` to get the preedit string, show it in your UI.

I'll make a screen record to show how text is inputted with fcitx5 in a normal GTK application.

**Comment by lilydjwg** [comment]
This is how an input method is in use. The purple window is from the input method, and the application window is `gtk4-demo-application`.

https://github.com/gabm/Satty/assets/440661/163a0caa-4db2-494b-aa10-c4a548b2fbee

You need to tell the input method that the user is going to input text, where the cursor is, and retrieve and draw both preedit and committed text on your UI. 

Although only the text tool needs this, it's an important annotation tool for screenshots, so I really hope it works for us.

**Comment by gabm** [comment]
I made progress on this. Would you be able to build from source and test?

**Comment by ystyle** [comment]
> I made progress on this. Would you be able to build from source and test?

I can try, but it will have to be tomorrow (it's night here now)

**Comment by gabm** [comment]
I'll let you know when i pushed it.. still some refactoring needed

**Comment by lilydjwg** [comment]
I can test too.

**Comment by gabm** [comment]
Hi both, thank you for volunteering. I pushed a branch `multi-input` that uses the `IMMulticontext`  as seen in https://github.com/flxzt/rnote/blob/main/crates/rnote-ui/src/canvas/mod.rs#L473. There is currently no "pre-edit" text handling because I don't know how to draw that currently...

Can you let me know if that works and maybe create a screen recording?

**Comment by gabm** [comment]
sorry to hear that.. how can i test this myself?

**Comment by lilydjwg** [comment]
Try to [install fcitx5](https://wiki.archlinux.org/title/Fcitx5)? You need to add and activate one input method (one for any language will do).

**Comment by Clareses** [comment]
Hello, I've submitted a PR for this issue. Please take a look. Thanks. @gabm #52

**Comment by xiaohuirong** [comment]
> Hello, I've submitted a PR for this issue. Please take a look. Thanks. @gabm #52

I tried your PR, and setting 'GTK_IM_MODULE=fcitx' under x11 and wayland works, but it does not support the text-input protocol for wayland. I encountered an issue where the input method dialog under wayland always appears at the top-left corner of one screen and does not follow the input position. Under Wayfire and Sway, it always appears on the left screen, while under Hyprland, it always appears on the right screen.

![screenshot resized](https://github.com/gabm/Satty/assets/69972645/4eb214c3-2a96-478e-b9a4-e44bfa5dd086)

**Comment by xiaohuirong** [comment]
In Wayland, using --gtk-version=4 to enable the GTK4 input method module support in Chromium also has a similar position offset issue. https://issues.chromium.org/issues/40929728

**Comment by lilydjwg** [comment]
I have another idea for this issue: overlay a transparent-background GtkTextView on where text is to be inputted. This can solve the input method issue as well as copy & paste and shortcut keys etc. Unfortunately I didn't figure out how to use relm4.

**Comment by lilydjwg** [comment]
Yes with `GTK_IM_MODULE=fcitx` it works to some extent for me too. Besides the popup window placement issue, the preedit string is also not displayed.

**Comment by gabm** [comment]
The first version got released as part of https://github.com/gabm/Satty/releases/tag/v0.10.0. Since there are still open issues, I will re-open this.

**Comment by lilydjwg** [comment]
No. It doesn't work with either text input or gtk module.

**Comment by Clareses** [comment]
I tried merging opengl branch into the multiinput branch and entered non-Latin alphabet characters. OpenGL seems to be working fine.
![2024-02-27-12-02-54](https://github.com/gabm/Satty/assets/79776530/c9f95858-f5a2-47cd-a72a-3560392bdf27)

**Comment by lilydjwg** [comment]
Oops, I didn't merge the multiinput branch. It has a lot of conflicts to merge.

**Comment by gabm** [comment]
oops, I assumed the opengl branch did have those patches already - sorry. 

Just to clarify: the main branch includes the multi-input patches and its released as v0.10.0. OpenGL will - once released - of course include all previous patches/features.

@Clareses thanks for merging and testing it - nice to see it works!

**Comment by rtgiskard** [comment]
currently with satty 0.12.0, I have to set `GTK_IM_MODULE=fcitx5` to make it works with fcitx5, however fcitx5 seems to deprecated `GTK_IM_MODULE`

**Comment by Guan9Wu** [comment]
with satty 0.13.0, can't input chinese by fcitx5 in Hyprland (a wayland-based desktop environment), and I set GTK_IM_MODULE=fcitx5, but still can't works.

**Comment by gabm** [comment]
@Clareses is there something you could do?

**Comment by wenjinnn** [comment]
Could this related to some breaking change in Hyprland? It dropped wlroots as dependency and replaced it with https://github.com/hyprwm/aquamarine, after version 0.42

**Comment by agata** [comment]
With my changes I can now enter Japanese via the IME at the start of a text layer. Tested in a Hyprland (Omarchy) environment.

<img width="1201" height="900" alt="Image" src="https://github.com/user-attachments/assets/85202e3c-b48b-4060-b2e1-9530ad180af2" />

I pushed these changes on https://github.com/agata/Satty/tree/improve-ime-support.

- load the configured font (when available) together with a stack of CJK fallback fonts inside the FemtoVG canvas, and share that stack with the text tool so glyphs from IME composition render correctly;
- wire gtk::IMMulticontext preedit/commit events through SketchBoard and keep the IME cursor rectangle in sync so GTK can anchor the candidate window at the right spot;
- teach the text tool to manage IME preedit strings, handle commit/cancel flows, and update the caret rectangle that we expose back to GTK.

Before launching Satty, export the IME variables so GTK and Qt hand off to fcitx:

```
export GTK_IM_MODULE=fcitx
export QT_IM_MODULE=fcitx
export XMODIFIERS=@im=fcitx
```

With those exports in place I can type Japanese through the IME end-to-end—the preedit string shows up and commits correctly, although the candidate window still doesn’t follow the on-canvas caret.

I’m not very familiar with Rust and mostly pieced this together “by vibe,” so I’m unsure whether it’s ready for a pull request; I’m very sorry, but if this fix turns out to be useful, I’d be grateful if someone more comfortable with Rust could help apply it.

**Comment by gabm** [comment]
@agata thats great! can you create a pull request with those changes for review?

**Comment by agata** [comment]
@gabm I've added #273 !

**Comment by RobertMueller2** [comment]
This should be working now in the latest release.

**Comment by zefr0x** [comment]
> This should be working now in the latest release.

Arabic is not working for me, it also has non-latin alphabet.
When I write it moves the text pointer, but letters doesn't appear.

Is there something that I should configure in my system?

**Comment by RobertMueller2** [comment]
> > This should be working now in the latest release.
> 
> Arabic is not working for me, it also has non-latin alphabet. When I write it moves the text pointer, but letters doesn't appear.
> 
> Is there something that I should configure in my system?

At this point, IME requires using a font that actually contains the glyphs that are being entered. We need to improve fallback font usage/configuration down the road, this is tracked in #291.

EDIT: Sorry, I really should have mentioned that above. :)

**Comment by lilydjwg** [comment]
> This should be working now in the latest release.

I just tried 9792b10 and fcitx5 doesn't activate in satty with either wayland or dbus protocol.

**Comment by RobertMueller2** [comment]
> > This should be working now in the latest release.
> 
> I just tried [9792b10](https://github.com/gabm/Satty/commit/9792b1016c5a04d5280abb6553dc2ba6c4aa9181) and fcitx5 doesn't activate in satty with either wayland or dbus protocol.

Which language are you using in fctix5, exactly? And what's your `GTK_IM_MODULE` environment set to? I'm thinking the latest change may have broke it for some languages, e.g. it does work for Japanese but not Chinese. Not sure yet why.

Just for clarification, can you use ctrl+space (or whichever hotkey you are using for fcitx5 activation) at all? Because that's working for me with all languages.

**Comment by agata** [comment]
@lilydjwg 

My change might have broken it.

85952ec

If you revert to the commit above, you can test without including my changes (005cf12, 8719b39). Could you please check if it works on your end?

**Comment by lilydjwg** [comment]
`git revert` those two commits fails so I tested the parent commit 85952ec52f74f7648a5f18f8a8626a55df4f8ab3 instead. It worked(to some extent: no font fallback was done so that I needed to specify a font family; the weight seemed to be slim even I specified Bold or Heavy styles (why not just using pango?) Also the position of input method window was very wrong.

**Comment by agata** [comment]
Thanks for the update, @lilydjwg.

You're right, I've switched to using Pango in #308. I tested it with Japanese, and I believe there have been significant improvements, particularly with caret positioning.

Could you please re-test and let me know your thoughts?

**Comment by Bryan2333** [comment]
I built satty on main branch, Fcitx5 can be activated in satty now. But the text is not visible.

https://github.com/user-attachments/assets/fcbf50fe-cf40-4f18-8af4-db29547708b4

**Comment by RobertMueller2** [comment]
> I built satty on main branch, Fcitx5 can be activated in satty now. But the text is not visible.
> _20251106_151004.webm

Very likely due to the configured font not containing the glyphs. We have #291, #298,  #309 for this.

Also, the README says, "Satty supports IME via GTK with and without preediting. Please note, at this point Satty has no proper fallback font handling so the font used needs to contain the entered glyphs."

**Comment by Bryan2333** [comment]
> > I built satty on main branch, Fcitx5 can be activated in satty now. But the text is not visible.
> > _20251106_151004.webm
> 
> Very likely due to the configured font not containing the glyphs. We have [#291](https://github.com/Satty-org/Satty/issues/291), [#298](https://github.com/Satty-org/Satty/pull/298), [#309](https://github.com/Satty-org/Satty/issues/309) for this.
> 
> Also, the README says, "Satty supports IME via GTK with and without preediting. Please note, at this point Satty has no proper fallback font handling so the font used needs to contain the entered glyphs."

After I specified the default font, the text is visible now. Thanks.

**Comment by RobertMueller2** [comment]
> > > I built satty on main branch, Fcitx5 can be activated in satty now. But the text is not visible.
> > > _20251106_151004.webm
> > 
> > 
> > Very likely due to the configured font not containing the glyphs. We have [#291](https://github.com/Satty-org/Satty/issues/291), [#298](https://github.com/Satty-org/Satty/pull/298), [#309](https://github.com/Satty-org/Satty/issues/309) for this.
> > Also, the README says, "Satty supports IME via GTK with and without preediting. Please note, at this point Satty has no proper fallback font handling so the font used needs to contain the entered glyphs."
> 
> After I specified the default font, the text is visible now. Thanks.

yeah, i'm sorry the user  experience is not better yet. But I'm hoping we can get #298 done for the next minor release, that should improve it a bit.

**Comment by RobertMueller2** [comment]
Anyone here able to have a look at #555, specifically if IME still works as before? @robertwidfen  and I have both tested it and think it's ok, but as we're not daily IME users, it'd be good to get some additional feedback. Thanks :)

**Comment by lilydjwg** [comment]
It works the same as main, but not before[^1]:

<img width="187" height="134" alt="Image" src="https://github.com/user-attachments/assets/1e691b04-d636-4a31-8fc2-660298358f6e" />

The input window is covering the bottom part of text. You should put it just below the box you draw around the text. It means that you report the exact text cursor shape. The top of reported cursor position is also a bit off, but since the cursor is higher that actual text, it doesn't cause any harm.

The underline is also too high. It is expected that the underline is over the bottom part of text, but not at this high.

<img width="142" height="78" alt="Image" src="https://github.com/user-attachments/assets/87852f61-fb7c-450b-89ea-bdadda92a656" />

[^1]: last time I tested, if I remember correctly

**Comment by RobertMueller2** [comment]
With #593 merged, contained in 0.22.0, are there any leftover issues here?

**Comment by lilydjwg** [comment]
The font fallback problem remains: I need to specify a font containing Han characters or I get hollow boxes. This means that I can't use characters from different fonts at the same time, e.g. emojis.

Also, when I try to input a character using the Unicode input support from fcitx5 (shortcut <kbd>Ctrl-Shift-Alt-u</kbd> and type to search for a Unicode character), the text may have a blue stroke. I'm not sure what happened but it seems to relate to key handling.

<img width="275" height="142" alt="Image" src="https://github.com/user-attachments/assets/1fb029a2-d407-4cdf-87bc-a404d47a6b30" />

**Comment by RobertMueller2** [comment]
> The font fallback problem remains: I need to specify a font containing Han characters or I get hollow boxes. This means that I can't use characters from different fonts at the same time, e.g. emojis.

Are you using the fallback config option? That should work, afaik.

_Automatic_ fallback font detection is tracked in #309.
 
> Also, when I try to input a character using the Unicode input support from fcitx5 (shortcut Ctrl-Shift-Alt-u and type to search for a Unicode character), the text may have a blue stroke. I'm not sure what happened but it seems to relate to key handling.
> <img alt="Image" width="275" height="142" src="https://private-user-images.githubusercontent.com/440661/655638344-1fb029a2-d407-4cdf-87bc-a404d47a6b30.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3ODk5ODIwNzYsIm5iZiI6MTc4OTk4MTc3NiwicGF0aCI6Ii80NDA2NjEvNjU1NjM4MzQ0LTFmYjAyOWEyLWQ0MDctNGNkZi04N2JjLWE0MDRkNDdhNmIzMC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwOTIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDkyMVQwOTA5MzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NGU0NTFkYzgzOGViMmZiNDQ4ZGFhODAwYjQ1ODQ5YWVkMWM1MDBjZWZiNjlhNWQ4ZDZkY2I5ODhkM2FlOWIwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZyZXNwb25zZS1jb250ZW50LXR5cGU9aW1hZ2UlMkZwbmcifQ.8vUeBNYSoxx8Te6FzxRnQ9IVO6G1LBqei7sBtJAqRy8">

This is outlining, triggered by alt. I assume the shortcut you mentioned is interpreted as triggering this. Which version are you using right now? We should open a bug for it if it still exists in the current repository head commit.

**Comment by lilydjwg** [comment]
Yes, the font fallback option works after I configured it. I hope it can be default to use the fallback list from fontconfig. My emoji font Twemoji (colrv1) doesn't work however.

I was using 0.22.0. I jut tried the latest commit (2bcd9111390a7ae03ef965e87a6a12dfd22bf93b) and the outline still triggers unexpectedly.
